### PR TITLE
Decorating with views in phoenix

### PR DESCRIPTION
**Pull request by danmcclain**

### CODE REVIEW

File: source/posts/2016-01-15-phoenix-best-practice-decorating-in-views.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+those mistakes with the following patterns.
+
+## Keep your database out of my view by preloading
+
+The view layer in Phoenix serves as a [decorator][decorator-pattern] for
+the template you are about to render. It serves to merge data retrieved
+in the controller with the additional details that may be needed. On our
+[JSON API][json-api] backends, we use [ja_serializer][ja_serializer] to
+convert our models into the proper response. When we have additional
+details (say we include a blog post's author for side loading into the
```
Comment (michaeldupuisjr) [inline source/posts/2016-01-15-phoenix-best-practice-decorating-in-views.md]: into Ember

File: source/posts/2016-01-15-phoenix-best-practice-decorating-in-views.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+Ember), we  want to make sure that we preload this data before it gets
+to the view. If we preload from the view itself, you will be creating an
```
Comment (michaeldupuisjr) [inline source/posts/2016-01-15-phoenix-best-practice-decorating-in-views.md]: itself, *we* will

**Comment by michaeldupuisjr** [comment]
Solid post @danmcclain!

File: source/posts/2016-01-15-phoenix-best-practice-decorating-in-views.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+[N+1 scenario][nplusone]. In the case of the index route, where you,
+for example, you may have 20 posts you want to render. When we wait until
```
Comment (brzpegasus) [inline source/posts/2016-01-15-phoenix-best-practice-decorating-in-views.md]: Something's off with this sentence:

> In the case of the index route, where you, for example, you may have 20 posts you want to render.

File: source/posts/2016-01-15-phoenix-best-practice-decorating-in-views.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+we get to the view's `show.json`, which renders the individual post, to
+retrieve the author, we will be doing 20 additional SQL queries, as we
+will be getting the post one by one. If we preload the author
+relationship in the controller, we will be do a single query that
```
Comment (brzpegasus) [inline source/posts/2016-01-15-phoenix-best-practice-decorating-in-views.md]: "we will be do" -> "we will be doing"

File: source/posts/2016-01-15-phoenix-best-practice-decorating-in-views.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+You may have additional information you want your view to have access
+to. In one case that I came across, we query the database to find the
+number of blog posts a specific tag has. We end up storing the
+information in a map in the function we created to look up that
+information, where the key is the id of the tag, and the value is the
+number of posts for that tag. We just went over why you don't want to do
+this work in your view, so how do we pass that information down? We can
+use the `Plug.Conn.assign` ([docs][assign-docs]) to store the
+information on the `conn` which is passed to render in the controller.
```
Comment (brzpegasus) [inline source/posts/2016-01-15-phoenix-best-practice-decorating-in-views.md]: Perhaps a code snippet might help too?

File: source/posts/2016-01-15-phoenix-best-practice-decorating-in-views.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+retrieval in the controller, it actually makes it quite difficult to
+create an N+1 query. When you have a list of records, the easiest way to
+preload data (if you aren't using `Ecto.Query.preload`), is to call
+`Repo.preload(collection, keys)`, which will only perform a single query
+in the end.
+
+Moving from Rails to Phoenix, I've personally found there to be fewer
+footguns to shoot myself with. I think a big part of that ends up being
+the fact that with a functional language, you can't accidentally make a
+call to the database because your relationship can load itself. With
```
Comment (brzpegasus) [inline source/posts/2016-01-15-phoenix-best-practice-decorating-in-views.md]: can -> can't?

**Pull request merged**